### PR TITLE
Filter stacktrace on newer NodeJS

### PR DESCRIPTION
**Pull request by HonzaMac**
NodeJS changed way how stacktrace look for async hooks:
Example:
```
    at AsyncHook.init (/www/node_modules/blocked-at/index.js:31:11)
    at TCP.emitInitNative (node:internal/async_hooks:201:43)
    at Socket.connect (node:net:993:7)
    at Object.connect (node:net:203:17)
    at Object.netConnect [as createConnection] (/www/node_modules/puppeteer/node_modules/ws/lib/websocket.js:1006:14)
    at new ClientRequest (node:_http_client:326:45)
    at request (node:http:96:10)
```

### CODE REVIEW

**Comment by HonzaMac** [comment]
Cc @naugtur ?

**Comment by naugtur** [comment]
UX fail, I gave you a review on the phone and now (desktop) I see it's pending...

File: index.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ const asyncHooks = require('async_hooks')
 const cache = new Map()// WeakMap maybe?
 let resourcesCount = 0
 
-const asyncHooksRegEx = /\((internal\/)?async_hooks\.js:/
+const asyncHooksRegEx = /\(((node:)?internal[:/])?async_hooks(\.js)?:/
```
[the remark this answers is not in the document]
Comment (naugtur) [reply index.js:7]: I wonder if this is proof enough https://redosdetector.com/?pattern=%5C%28%28%28node%3A%29%3Finternal%5B%3A%2F%5D%29%3Fasync_hooks%28%5C.js%29%3F%3A&unicode=false

[the remark this answers is not in the document]
Comment (HonzaMac) [reply index.js]: I have split those regexes, so they can be dropped with nodejs version support.

File: index.js [hunk 2]
```diff
@@ -4,13 +4,18 @@ const asyncHooks = require('async_hooks')
 const cache = new Map()// WeakMap maybe?
 let resourcesCount = 0
 
-const asyncHooksRegEx = /\((internal\/)?async_hooks\.js:/
+const asyncHooksRegExNode8 = /\((internal\/)?async_hooks\.js:/
+const asyncHooksRegExNode16 = /node:internal\/async_hooks:/
+
+function isInternalStacktrace (frames, i) {
+  return !asyncHooksRegExNode16.test(frames[i]) && !asyncHooksRegExNode8.test(frames[i])
```
Comment (naugtur) [inline index.js:11]: I like it that new versions don't pay the penalty for older version support.

File: index.js [hunk 3]
```diff
@@ -4,13 +4,18 @@ const asyncHooks = require('async_hooks')
...
+}
 
 function cleanStack (stack) {
   const frames = stack.split('\n')
   // this part is opinionated, but it's here to avoid confusing people with internals
   let i = frames.length - 1
-  while (i && !asyncHooksRegEx.test(frames[i])) {
+  while (i && isInternalStacktrace(frames, i)) {
```
Comment (naugtur) [inline index.js:18]: What's the point of this change?

File: test/index.js [hunk 4]
```diff
@@ -26,8 +26,8 @@ setTimeout(_ => {
   assert(!result.includes('Error'), 'Expected Error line not to be included')
   const stacks = result.join()
   assert(!stacks.includes('async_hooks.js'), 'Expected async_hooks.js traces not to be included')
+  assert(!stacks.includes('async_hooks'), 'Expected async_hooks traces not to be included')
   assert(stacks.includes('test' + sep + 'index.js:18'), 'Expected test' + sep + 'index.js:18 in stack trace')
-  assert(stacks.includes('Server.connectionListener'), 'Expected Server.connectionListener in stack trace')
```
Comment (naugtur) [inline test/index.js:30]: Is there no useful information about going through the server implementation in the stacktrace anymore?

**Comment by naugtur** [comment]
Last round of questions, I'll merge and release once we finish talking :)

File: test/index.js (hunk 4 above)
Comment (HonzaMac) [reply test/index.js:30]: Seems there are only internal timeouts mentioned in stacktrace for this test case.

Comment (naugtur) [reply test/index.js:30]: Oh that's sad. It'd be useful to the end user.

**Comment by sraka1** [comment]
@naugtur this doesn't seem like it was ever release to npm?

**Pull request merged**